File: game/xiangqi/logic/bot.ts
import {
  EMPTY,
  KIND_ADVISOR,
  KIND_CANNON,
  KIND_CHARIOT,
  KIND_ELEPHANT,
  KIND_GENERAL,
  KIND_HORSE,
  KIND_SOLDIER,
  boardX,
  boardY,
  crossedRiver,
  pieceKind,
  pieceSide,
} from './board';
import { inCheck, legalMovesFrom, terminalKind } from './moves';

export type BotDifficulty = 'easy' | 'medium' | 'hard';

export interface BotMove {
  from: number;
  to: number;
}

export const BOT_DIFFICULTY_LABEL: Record<BotDifficulty, string> = {
  easy: 'Dễ',
  medium: 'Vừa',
  hard: 'Khó',
};

const PIECE_VALUE: Record<number, number> = {
  [KIND_GENERAL]: 100_000,
  [KIND_ADVISOR]: 210,
  [KIND_ELEPHANT]: 210,
  [KIND_HORSE]: 420,
  [KIND_CHARIOT]: 900,
  [KIND_CANNON]: 450,
  [KIND_SOLDIER]: 100,
};

const MATE_SCORE = 1_000_000;

export function legalMovesForSide(board: readonly number[], side: number): BotMove[] {
  const moves: BotMove[] = [];
  for (let from = 0; from < board.length; from++) {
    const piece = board[from];
    if (piece === EMPTY || pieceSide(piece) !== side) continue;
    for (const to of legalMovesFrom(board, from)) moves.push({ from, to });
  }
  return moves;
}

function afterMove(board: readonly number[], move: BotMove): number[] {
  const next = [...board];
  next[move.to] = next[move.from];
  next[move.from] = EMPTY;
  return next;
}

function pieceSquareBonus(piece: number, square: number): number {
  const side = pieceSide(piece);
  const kind = pieceKind(piece);
  const x = boardX(square);
  const y = boardY(square);
  if (kind === KIND_SOLDIER) {
    const advance = side === 0 ? y : 9 - y;
    return advance * 8 + (crossedRiver(y, side) ? 36 + (4 - Math.abs(4 - x)) * 4 : 0);
  }
  if (kind === KIND_HORSE || kind === KIND_CANNON) return (4 - Math.abs(4 - x)) * 4;
  if (kind === KIND_CHARIOT) return (4 - Math.abs(4 - x)) * 2;
  return 0;
}

function evaluate(board: readonly number[], perspective: number): number {
  let score = 0;
  for (let square = 0; square < board.length; square++) {
    const piece = board[square];
    if (piece === EMPTY) continue;
    const value = PIECE_VALUE[pieceKind(piece)] + pieceSquareBonus(piece, square);
    score += pieceSide(piece) === perspective ? value : -value;
  }
  if (inCheck(board, 1 - perspective)) score += 28;
  if (inCheck(board, perspective)) score -= 28;
  return score;
}

function movePriority(board: readonly number[], move: BotMove): number {
  const captured = board[move.to];
  const mover = board[move.from];
  let score = captured === EMPTY ? 0 : PIECE_VALUE[pieceKind(captured)] * 10 - PIECE_VALUE[pieceKind(mover)];
  const next = afterMove(board, move);
  if (inCheck(next, 1 - pieceSide(mover))) score += 180;
  return score;
}

function orderedMoves(board: readonly number[], side: number): BotMove[] {
  return legalMovesForSide(board, side).sort((a, b) => movePriority(board, b) - movePriority(board, a));
}

interface SearchContext {
  perspective: number;
  deadline: number;
  nodes: number;
}

function search(
  board: readonly number[],
  side: number,
  depth: number,
  alpha: number,
  beta: number,
  ply: number,
  context: SearchContext,
): number {
  context.nodes++;
  if ((context.nodes & 127) === 0 && Date.now() >= context.deadline) return evaluate(board, context.perspective);
  const terminal = terminalKind(board, side);
  if (terminal !== 'none') return side === context.perspective ? -MATE_SCORE + ply : MATE_SCORE - ply;
  if (depth === 0) return evaluate(board, context.perspective);

  const maximizing = side === context.perspective;
  let best = maximizing ? -Infinity : Infinity;
  for (const move of orderedMoves(board, side)) {
    const score = search(afterMove(board, move), 1 - side, depth - 1, alpha, beta, ply + 1, context);
    if (maximizing) {
      best = Math.max(best, score);
      alpha = Math.max(alpha, best);
    } else {
      best = Math.min(best, score);
      beta = Math.min(beta, best);
    }
    if (beta <= alpha || Date.now() >= context.deadline) break;
  }
  return best;
}

export function chooseBotMove(
  board: readonly number[],
  side: number,
  difficulty: BotDifficulty,
  random: () => number = Math.random,
): BotMove | null {
  const moves = orderedMoves(board, side);
  if (moves.length === 0) return null;
  if (difficulty === 'easy') return moves[Math.min(moves.length - 1, Math.floor(random() * moves.length))];

  const depth = difficulty === 'hard' ? 3 : 2;
  const budgetMs = difficulty === 'hard' ? 650 : 220;
  const context: SearchContext = { perspective: side, deadline: Date.now() + budgetMs, nodes: 0 };
  const scored = moves.map((move) => ({
    move,
    score: search(afterMove(board, move), 1 - side, depth - 1, -Infinity, Infinity, 1, context),
  }));
  scored.sort((a, b) => b.score - a.score || movePriority(board, b.move) - movePriority(board, a.move));

  if (difficulty === 'medium') {
    const candidates = scored.filter((item) => item.score >= scored[0].score - 45).slice(0, 3);
    return candidates[Math.min(candidates.length - 1, Math.floor(random() * candidates.length))].move;
  }
  return scored[0].move;
}

File: server/internal/game/games/xiangqi/bot.go
package xiangqi

import (
	"sort"
	"time"
)

const (
	mateScore = 1_000_000
	infScore  = 1 << 30

	timeoutMoveDepth  = 3
	timeoutMoveBudget = 300 * time.Millisecond
)

var pieceValue = func() [MaxPiece + 1]int {
	byKind := [KindSoldier + 1]int{
		KindGeneral:  100_000,
		KindAdvisor:  210,
		KindElephant: 210,
		KindHorse:    420,
		KindChariot:  900,
		KindCannon:   450,
		KindSoldier:  100,
	}
	var table [MaxPiece + 1]int
	for piece := 1; piece <= MaxPiece; piece++ {
		table[piece] = byKind[pieceKind(piece)]
	}
	return table
}()

type botMove struct {
	From int
	To   int
}

func legalMovesForSide(board []int, side int) []botMove {
	var moves []botMove
	for from := 0; from < len(board); from++ {
		piece := board[from]
		if piece == Empty || pieceSide(piece) != side {
			continue
		}
		for _, to := range legalMovesFrom(board, from) {
			moves = append(moves, botMove{From: from, To: to})
		}
	}
	return moves
}

func afterMove(board []int, move botMove) []int {
	next := make([]int, len(board))
	copy(next, board)
	next[move.To] = next[move.From]
	next[move.From] = Empty
	return next
}

func centerBonus(x int) int {
	d := 4 - x
	if d < 0 {
		d = -d
	}
	return 4 - d
}

func pieceSquareBonus(piece, square int) int {
	side := pieceSide(piece)
	x, y := boardX(square), boardY(square)
	switch pieceKind(piece) {
	case KindSoldier:
		advance := y
		if side != SideRed {
			advance = BoardHeight - 1 - y
		}
		score := advance * 8
		if crossedRiver(y, side) {
			score += 36 + centerBonus(x)*4
		}
		return score
	case KindHorse, KindCannon:
		return centerBonus(x) * 4
	case KindChariot:
		return centerBonus(x) * 2
	}
	return 0
}

func evaluate(board []int, perspective int) int {
	score := 0
	for square, piece := range board {
		if piece == Empty {
			continue
		}
		value := pieceValue[piece] + pieceSquareBonus(piece, square)
		if pieceSide(piece) == perspective {
			score += value
		} else {
			score -= value
		}
	}
	if inCheck(board, 1-perspective) {
		score += 28
	}
	if inCheck(board, perspective) {
		score -= 28
	}
	return score
}

func movePriority(board []int, move botMove) int {
	captured := board[move.To]
	mover := board[move.From]
	score := 0
	if captured != Empty {
		score = pieceValue[captured]*10 - pieceValue[mover]
	}
	if inCheck(afterMove(board, move), 1-pieceSide(mover)) {
		score += 180
	}
	return score
}

type ratedMove struct {
	Move     botMove
	Priority int
}

func orderedMoves(board []int, side int) []ratedMove {
	moves := legalMovesForSide(board, side)
	rated := make([]ratedMove, len(moves))
	for i, move := range moves {
		rated[i] = ratedMove{Move: move, Priority: movePriority(board, move)}
	}
	sort.SliceStable(rated, func(i, j int) bool {
		return rated[i].Priority > rated[j].Priority
	})
	return rated
}

type searchContext struct {
	perspective int
	deadline    time.Time
	nodes       int
}

func (c *searchContext) outOfTime() bool {
	return !time.Now().Before(c.deadline)
}

func search(board []int, side, depth, alpha, beta, ply int, context *searchContext) int {
	context.nodes++
	if context.nodes&127 == 0 && context.outOfTime() {
		return evaluate(board, context.perspective)
	}
	if !hasLegalMove(board, side) {
		if side == context.perspective {
			return -mateScore + ply
		}
		return mateScore - ply
	}
	if depth == 0 {
		return evaluate(board, context.perspective)
	}

	maximizing := side == context.perspective
	best := infScore
	if maximizing {
		best = -infScore
	}
	for _, rated := range orderedMoves(board, side) {
		score := search(afterMove(board, rated.Move), 1-side, depth-1, alpha, beta, ply+1, context)
		if maximizing {
			if score > best {
				best = score
			}
			if best > alpha {
				alpha = best
			}
		} else {
			if score < best {
				best = score
			}
			if best < beta {
				beta = best
			}
		}
		if beta <= alpha || context.outOfTime() {
			break
		}
	}
	return best
}

// chooseBestMove mirrors the "hard" difficulty of the client bot in
// game/xiangqi/logic/bot.ts, minus the randomness the client uses to vary
// easy and medium play.
func chooseBestMove(board []int, side int, depth int, budget time.Duration) (botMove, bool) {
	moves := orderedMoves(board, side)
	if len(moves) == 0 {
		return botMove{}, false
	}
	context := &searchContext{perspective: side, deadline: time.Now().Add(budget)}
	scores := make([]int, len(moves))
	for i, rated := range moves {
		scores[i] = search(afterMove(board, rated.Move), 1-side, depth-1, -infScore, infScore, 1, context)
	}
	bestIdx := 0
	for i := 1; i < len(moves); i++ {
		if scores[i] > scores[bestIdx] ||
			(scores[i] == scores[bestIdx] && moves[i].Priority > moves[bestIdx].Priority) {
			bestIdx = i
		}
	}
	return moves[bestIdx].Move, true
}
